### PR TITLE
Optimalize eagle text generation techniques

### PR DESCRIPTION
**Pull request by michalspiegel**
I manually evaluated the quality of the generated texts on czech, slovak and english languages and tried to improve the methods with different prompts and model generation parameters.

Main findings:
- Paraphrasing seems to work in the Instruction:/Input:/Response: - It seems to have something to do with how the model was trained. (It sometimes includes this format even if it wasn't in the prompt)
- Paraphrasing using other prompts - the model often misunderstands the prompt
- Playing with different model generation parameters doesn't seem to do much - other than temperature - low temperature and the models starts to repeat itself often - higher temperature is better for more "diverse" answer
- For keywords method, again, using the Instruction:/Response: prompt template seems best
- Ending with "Response:" is best, the model less often continues in generating the instruction or other unwanted text
- With k_to_one, the model is often very confused from such complex instruction
- With k_to_one it seems its best to just provide the different social post examples in the prompt directly without any instruction and just let him continue generating
- I tried this "no instruction" approach also with paraphrasing but it sometimes generates gibberish like:
```
############## Prompt ##############
{'min_new_tokens': 0, 'max_new_tokens': 100, 'num_return_sequences': 1, 'do_sample': True, 'num_beams': 1, 'top_k': 50, 'top_p': 0.95, 'temperature': 1}
are gays allowed in budhism
cause ngl buddha sound kinda lit
Id fuck with him

############## Response ##############
are gays allowed in budhism
cause ngl buddha sound kinda lit
Id fuck with him
--- 897650105
>>897648948
--- 897650342
>>897648948
--- 897650420
>>897650269
--- 897650453
>>897648948
--- 897650522
>>897650355
--- 897650548
>>897648948
--- 897650575
>>897650453
nice
```
- This doesn't seem to happen when the text is longer (e.g. multiple posts in the k_to_one approach). The model seems to be more grounded and generates the correct format

### CODE REVIEW

File: eagle.py [hunk 1]
```diff
@@ -18,47 +55,44 @@ def __init__(self, model_name: str,  debug: bool = False, use_gpu: bool = False,
...
-            text = self.query(prompt)
-        return text
+        for template in PROMPTS:
+            for config in PARAMETERS:
+                prompt = template.format(language=LANGUAGE_CODE_MAP[language], text=text)
+                #instruction = f"Your goal is to paraphrase text in {LANGUAGE_CODE_MAP[language]} using different words and sentence composition. Respond with only paraphrased text and nothing else. Text to paraphrase:"
+                for _ in range(iterations):
+                    #prompt = f'{instruction} "{text}"'
+                    answer = self.query(prompt, MODEL_GENERATE_ARGS=config)
+        return answer
```
Comment (kopalja) [inline eagle.py:84]: I don't think this function does what you intended to do. it returns answer for the last `template` from `PROMPTS` and  `config` from `PARAMETERS`. All previous `answers` are lost and never used.

**Comment by kopalja** [comment]
@michalspiegel Thank you for your `Eagle` improvements. I just found one suspicious thing worth fixing regarding `model.paraphrase`.

**Comment by kopalja** [comment]
> * With k_to_one it seems its best to just provide the different social post examples in the prompt directly without any instruction and just let him continue generating

Is this really the case? Shouldn't we at lease specify the language we want to generate post in?

File: eagle.py (hunk 1 above)
Comment (michalspiegel) [reply eagle.py:84]: You're right. I'm aware of this. This is still a work in progress and more of a playground than a final product. For my purpose, it was enough to have the debug logs/messages which is what I had for all of them.

**Comment by michalspiegel** [comment]
> > * With k_to_one it seems its best to just provide the different social post examples in the prompt directly without any instruction and just let him continue generating
> 
> Is this really the case? Shouldn't we at lease specify the language we want to generate post in?

We don't need to specify the language since the text provided as prompt is already in that language, so the model is naturally compelled to generate further in this language.

**Pull request merged**